### PR TITLE
Demote update

### PR DESCRIPTION
**Pull request by lukenamop**
This adds !demote in src/comment_worker.py and src/message.py

It also addresses the issue of promoting an Executive when there is already a CFO (promotes them straight to COO now), and does the same with the reciprocal demotions.

Tests all new and changed !promote and !demote inputs in test/firms.py

Also addresses a few unrelated issues in test/invest.py and test/active.py

### CODE REVIEW

**Review by thecsw — COMMENTED** [review]
I took a quick glance at the changes, seems fine! Don't forget to add `!demote` command to the list of commands in `!help`. Expect this to be reviewed within several hours/days.

**Comment by lukenamop** [comment]
@thecsw Good catch, I added that in 👍

**Comment by lukenamop** [comment]
@thecsw I also was informed last night that '!tax' was not included in the help command, so I added that this morning.

File: src/comment_worker.py [hunk 1]
```diff
@@ -685,7 +694,93 @@ def promote(self, sess, comment, investor, to_promote):
...
+        if (user is None) or (user.name == investor.name) or (user.firm != investor.firm):
+            return comment.reply_wrap(message.demote_failure_org)
+
+        firm = sess.query(Firm).\
+            filter(Firm.id == user.firm).\
+            first()
+
+        user_role = user.firm_role
+
+        if user_role == "":
```
Comment (thecsw) [inline src/comment_worker.py:718]: Unnecessary "elif" after "return" [no-else-return]

There is only a return statement in this branch.

File: src/comment_worker.py [hunk 2]
```diff
@@ -685,7 +694,93 @@ def promote(self, sess, comment, investor, to_promote):
...
+            return comment.reply_wrap(message.demote_failure_trader_org)
+        elif user_role == "assoc":
```
Comment (thecsw) [inline src/comment_worker.py]: Line 718, 725, 737, 749 you have `elif` statements whan you can have just a simple `if` statement.

By the logic, you don't have a final set of instructions to execute, so all `elif`s can be changed into if.

File: src/comment_worker.py [hunk 3]
```diff
@@ -685,7 +694,93 @@ def promote(self, sess, comment, investor, to_promote):
                 REDDIT.subreddit(subreddit).flair.set(user.name, f"{firm.name} | {flair_role_user}")
                 REDDIT.subreddit(subreddit).flair.set(investor.name, f"{firm.name} | {flair_role_investor}")
 
-        return comment.reply_wrap(message.modify_promote(user))
+        return comment.reply_wrap(message.modify_promote(user, user_role))
+
+    @req_user
+    def demote(self, sess, comment, investor, to_demote):
```
Comment (thecsw) [inline src/comment_worker.py:701]: Can you also add some comments in your code? I am not asking for docstrings or official Sphinx kind of documentation. Just inline comments. For example,

``` python
if user_role == "":
 ...
```

Can be turned into 

``` python
# If investor is already at the lowest rank (floor trader), we can't demote him.
if user_role == "":
 ...
```

And something like that. Currently, I plan to sprinkle the code with inline commenting and docstrings so we can start forming an official documentation for the code. Thanks

File: src/comment_worker.py [hunk 4]
```diff
@@ -685,7 +694,93 @@ def promote(self, sess, comment, investor, to_promote):
...
+            max_assocs = max_assocs_for_rank(firm.rank)
+            if firm.assocs >= max_assocs:
+                return comment.reply_wrap(message.modify_demote_assocs_full(firm))
+
+            user.firm_role = "assoc"
+            firm.execs -= 1
+            firm.assocs += 1
+
+        elif user.firm_role == "cfo":
+            if investor.firm_role != "ceo":
```
Comment (thecsw) [inline src/comment_worker.py:741]: Yeah, just refer to the previous comment about in-code documentation.

File: src/comment_worker.py [hunk 5]
```diff
@@ -685,7 +694,93 @@ def promote(self, sess, comment, investor, to_promote):
...
+                firm.coo -= 1
+                firm.execs += 1
+            else:
+                user.firm_role = "cfo"
+                firm.coo -= 1
+                firm.cfo += 1
+
+        # Updating the flair in subreddits
+        flair_role_user = ''
+        if user.firm_role == "ceo":
```
Comment (thecsw) [inline src/comment_worker.py:772]: Probably all `elif`s here can be changed into simple `if`s

File: src/message.py [hunk 6]
```diff
@@ -240,7 +242,7 @@ def modify_active(active_investments):
         replace("%INVESTMENTS_LIST%", investments_list)
 
 MIN_INVEST_ORG = """
-The minimum possible investment is %MIN% MemeCoins (1% of your balance) or 100 memecoins. Whatever is higher.
+The minimum possible investment is %MIN% MemeCoins (1% of your balance) or 100 memecoins, whatever is higher.
```
Comment (thecsw) [inline src/message.py:245]: Good catch!

File: test/firms.py [hunk 7]
```diff
@@ -350,6 +350,16 @@ def test_wrong_firm(self):
         self.assertEqual(len(replies), 1)
         self.assertEqual(replies[0], message.promote_failure_org)
 
+    def test_promote_self(self):
```
Comment (thecsw) [inline test/firms.py:353]: I like your testing habits. Keep it up!

**Review by thecsw — CHANGES_REQUESTED** [review]
I left some comments in the code about logical statements and commenting. Please feel free to dispute any of those. Thanks

File: src/comment_worker.py (hunk 3 above)
Comment (thecsw) [reply src/comment_worker.py:701]: Also yeah, depending on your IDE, if you have a big/long comment, break it down to several lines. Usually every comment line is no longer than 80 characters.

Comment (thecsw) [reply src/comment_worker.py:701]:
``` python
# This is a multiline comment. This is a very big comment that looks weird so
# I have to break it down. Thank you for your attention.

**Comment by lukenamop** [comment]
Working on the requested changes now! Finally got some free time.

**Comment by lukenamop** [comment]
@thecsw Okay, I've made the changes and added comments to my code :)

**Review by thecsw — APPROVED** [review]
Tested and it looks good to me. Good job! Approved.

**Pull request merged**